### PR TITLE
New loader: generate dataframe

### LINKED ISSUE #519: Generate dataframe for new loader
## Context
This issue is part of the loader refactoring project to include microscopy data (#304) and to migrate to HDF5 loader (#474).

Following this [comment](https://github.com/ivadomed/ivadomed/issues/474#issuecomment-708688892), the first step is to generate a dataframe with filenames and images infos by checking data/derivatives availability and associated metadata.

It also aims to remove the dependency to bids_neuropoly.

## Steps to generate dataframe

1. Get a list of all files in BIDS folders starting with `sub*` and ending with specified extensions, including derivatives (ignore json or other files). Questions:
    * Should the extensions be specified as: all extensions compliant with BIDS for supported data_type (anat and microscopy) or extensions specified by user in the config file?
    * Should derivatives include: all files or only files with specified `seg-<label>` (target_suffix) in config file?

2. Parse all entities in filename (`sub`, `ses`, `sample`, `run`, `ce`, `down`, `acq`, `rec`) for supported modalities + `suffix` for modality + `seg-<label>` for derivatives

3. Create a dataframe with 1 line for each file with: `filename`, `all entities’ labels`, `modality`, `data_type`, `filetype`, `seg-<label>`, `absolute_path`, `dataset_root`

4. Read TSV files if present (`participants`, `samples`, `sessions`, `scans`), add columns to dataframe.

5. Read JSON sidecar files, add columns to dataframe.

6. Read appropriate metadata directly from file, depending on the filetype, add columns to dataframe.

## Other considerations

Somewhere, we should check derivatives availability. Currently, the check is done after the split when loading the dataset, both in traditional and adaptative loader. What is best?

We might want to take a look at or use [pybids](https://github.com/bids-standard/pybids) which is a package maintained by BIDS to interact with BIDS datsets.

## Next step

Once the dataframe is done, the next step would be the split which could be done by filtering one or several columns of the dataframe.

### PR DESCRIPTION
**Pull request by mariehbourget**
Fixes #519 
This PR is part of the loader refactoring project to include microscopy data (#304) and to migrate to HDF5 loader (#474).
It may also include changes to support EEG modality (#525).

It aims to generate a dataframe by indexing filenames and metadata in a bids path.

Changes have been tested with `anat` and `microscopy` datasets.

The changes are:
1 new function `create_bids_dataframe` in `ivadomed/loader/utils.py`.
1 new requirement: `pybids>=0.12.4`
3 new files:
`ivadomed/config/config_bids.json` : custom bids_config file needed for dealing with additional microscopy entities in `pybids`.
`ivadomed/config/config_new_loader.json` : ivadomed config file with 2 new `loader parameters`:  `bids_config` and `extensions`.
`dev/df_new_loader.py` : Script used to test the new `create_bids_dataframe` function.

I left a lot of comments in the code to explain the different steps and choices made.

I was able to keep the validation (by the BIDS-validator) by forcing the indexing of files in `sub-*` folders at the root of the bids path and of files specific to microscopy (`samples.tsv`, `samples.json`).

To discuss:
- Do we want to index derivatives? (depends on where and when we want to check for availability and do the dataset split)
- Currently, the files are indexed and filtered by `extensions` and `target_suffix`. We could also filter by `data_type` (anat, microscopy, etc) and/or modality suffix (ex: T1w, SEM, etc) in case multiple data type are available in a dataset.

TODO:
- Include parsing of `sessions.tsv` and `scans.tsv` files (not done at the moment, only `participants.tsv` and `samples.tsv`)
- Read appropriate metadata from image file and add to dataframe. Is it pertinent at this stage? Could also be done after the split when loading images to numpy arrays.
- Evaluate changes to make to include EEG.

### CODE REVIEW

File: ivadomed/config/config_new_loader.json
Comment (olix86) [inline ivadomed/config/config_new_loader.json:12]: same

File: ivadomed/config/config_new_loader.json
Comment (olix86) [inline ivadomed/config/config_new_loader.json:4]: I'd use a relative path such as
``` "log_directory": "log_directory/new_loader",```

that way it's easier for others to use it

File: ivadomed/loader/utils.py
Comment (olix86) [inline ivadomed/loader/utils.py:565]: Maybe consider [os.join](https://www.geeksforgeeks.org/python-os-path-join-method/)

File: ivadomed/loader/utils.py
Comment (andreanne-lemay) [inline ivadomed/loader/utils.py:512]: This is great :100: ! It makes it flexible for the dataset where we don't have derivatives. Concerning your question, I think we should index the derivatives if this option is set to True, it will be easier for the loading if everything is in the df. This function should also verify availability (in my opinion), so if we want to index derivatives, only the subjects with the right target(s) and contrast(s) should be in the final df.

File: ivadomed/loader/utils.py
Comment (andreanne-lemay) [inline ivadomed/loader/utils.py:532]: I understand why the indexing of 'samples.tsv' and 'samples.json' are forced, but why do we have to force the indexing of folders starting by "sub" won't they already be indexed by pybids?

File: ivadomed/loader/utils.py
Comment (andreanne-lemay) [inline ivadomed/loader/utils.py:535]: I see we have a fork of pybids, do we plan on overwriting some of their code? If so, we might want to handle the error when .json files exist, but are empty (https://github.com/ivadomed/ivadomed/issues/532). Currently, we still have an error in this case.

File: ivadomed/loader/utils.py
Comment (andreanne-lemay) [inline ivadomed/loader/utils.py:583]: I tested it on MRI data and it worked very well! I love the final dataframe format!

File: ivadomed/loader/utils.py
Comment (andreanne-lemay) [inline ivadomed/loader/utils.py:520]: We should make this parameter optional in the config file (we could simply put None by default) and if it is None, we take the default file present in ivadomed/ivadomed/config, because in most cases we don't need a custom bids file.

**Review by andreanne-lemay — COMMENTED** [review]
:1st_place_medal:

File: ivadomed/loader/utils.py
Comment (mariehbourget) [reply ivadomed/loader/utils.py:512]: Ok thanks, I also think it's better to check availability before the split, I'll look into that.

File: ivadomed/config/config_new_loader.json
Comment (mariehbourget) [reply ivadomed/config/config_new_loader.json]: Fixed in e5eb063

File: ivadomed/config/config_new_loader.json
Comment (mariehbourget) [reply ivadomed/config/config_new_loader.json]: Fixed in e5eb063

File: ivadomed/loader/utils.py
Comment (mariehbourget) [reply ivadomed/loader/utils.py]: Fixed in 7c10596

File: ivadomed/loader/utils.py
Comment (mariehbourget) [reply ivadomed/loader/utils.py]: Fixed in 4e5eaf7
Let me know if this is what you had in mind.

File: ivadomed/loader/utils.py
Comment (mariehbourget) [reply ivadomed/loader/utils.py:532]: This is because I kept the BIDS-validator "on", so files that are not BIDS-compliant are ignored. For microscopy, that means that sub folders at the root of the dataset are ignored because png and ometiff files are not compliant yet. So forcing "sub" is overkill for anat, but necessary for microscopy. When microscopy will be included in pybids and BIDS-validator, we will be able to remove the force_index.
This is not full proof because if non-compliant files are in the sub folders, they will be indexed to. I'll check if I can find a better condition.

File: ivadomed/loader/utils.py
Comment (mariehbourget) [reply ivadomed/loader/utils.py:535]: I was planning on making PR from the fork before I found the custom config file. For now, it seems easier to wait until BEP microscopy is merged before submitting changes to pybids. But the json file error has nothing to do with microscopy so yes, we could propose changes.

File: ivadomed/loader/utils.py
Comment (mariehbourget) [reply ivadomed/loader/utils.py:580]: Thanks!

File: ivadomed/loader/utils.py
Comment (jcohenadad) [inline ivadomed/loader/utils.py:510]: PEP8 asks for 2 lines between function defs

Comment (mariehbourget) [reply ivadomed/loader/utils.py:510]: Fixed in 85537da

File: ivadomed/loader/utils.py
Comment (mariehbourget) [reply ivadomed/loader/utils.py:517]: Filtering based on derivatives availability is done in b3c9d34. We now have to determine what is the behavior if no matching derivatives are found (currently raise an error and exit).

**Comment by mariehbourget** [comment]
Update in implementation

DONE:
-	Added filtering based on modality suffix (contrast)
-	Added filtering based on available derivatives
-	Added parsing or `sessions.tsv` and` scans.tsv` files

TODO:
-	Find a better condition for force index of microscopy files
-	Determine behavior if no matching derivatives are found (currently raise error and exit)
-	Load metadata from files (I'm not sure if it is the best place, should we do it in the loader specific to each file format instead?)
-	Create a test dataset including : `microscopy` (png + ometiff), `anat` (2D + 3D nifty), `dwi` (4D nifty), `eeg` (files TBD)
-	Adapt behavior of `scans.tsv` parsing for `eeg` and `dwi`.
-	Write a test for the function

File: ivadomed/loader/utils.py
Comment (jcohenadad) [inline ivadomed/loader/utils.py:614]: as discussed in meeting, you could inform with logger.debug if one derivative for a file is missing

File: ivadomed/loader/utils.py
Comment (mariehbourget) [reply ivadomed/loader/utils.py]: I updated the condition in 12a5b50. The indexing is now forced for sub-folders only if they include microscopy files based on extensions.

File: ivadomed/loader/utils.py
Comment (mariehbourget) [reply ivadomed/loader/utils.py:621]: Done in 7c3e8a2: warnings for missing one derivative for a subject, and missing all derivatives for a subject.

**Comment by mariehbourget** [comment]
This PR is ready for review.

As discussed:
- force index is now only applied to microscopy files
- if debugging is true: there are warnings for missing one derivative or all derivatives for a subject.

File: ivadomed/config/config_new_loader.json
Comment (charleygros) [inline ivadomed/config/config_new_loader.json:1]: I am guessing we will use this file during the prototyping / development of the new loader before removing it when fully integrated. Is it the idea?

File: ivadomed/config/config_new_loader.json
Comment (charleygros) [inline ivadomed/config/config_new_loader.json:13]: Does it mean this config_bids could potentially change across projects?

File: dev/df_new_loader.py
Comment (charleygros) [inline dev/df_new_loader.py:1]: Very well documented and useful. I am wondering if we should not move these lines to a proper testing function!

**Review by charleygros — APPROVED** [review]
Really well done @mariehbourget ! That was not a easy task to integrate this new hierarchy!
A small comment would be: to add some documentation about this new Galaxy in ReadTheDocs (e.g., re the new params in config file).

**Review by charleygros — COMMENTED** [review]
I appreciated how well your code was commented / documented!
Can I help with something? eg run some tests? any idea?

File: ivadomed/config/config_new_loader.json
Comment (mariehbourget) [reply ivadomed/config/config_new_loader.json:13]: Not really. This custom config is a modification to the default config of `pybids` to include microscopy data, since microscopy won't be merged with pybids for a while. It includes everything that is already in the default pybids config file + relevant entities for microscopy. It could be updated if microscopy BEP changes, or if we want to include other "non-bids" modality.
For MRI and any other "official" bids modality, the custom pybids config file is not necessary and should not be used.

File: dev/df_new_loader.py
Comment (mariehbourget) [reply dev/df_new_loader.py:1]: Yes absolutely! The purpose of this script was to test during development. I'll open another PR to include a proper testing function.

**Comment by mariehbourget** [comment]
> Really well done @mariehbourget ! That was not a easy task to integrate this new hierarchy!
> A small comment would be: to add some documentation about this new Galaxy in ReadTheDocs (e.g., re the new params in config file).

Thank you @charleygros for your feedback. I added the new params in the config file. I also added the key "extensions" to the default config and modify how the filtering is done with the extensions in the function. That way, "extensions" is not mandatory for MRI (as it was before), and is necessary only if a specific dataset contains more than one possible file extensions (e.g. microscopy with png and ometiff). Could you review the changes? Thanks!

I was wondering if it is OK to merge the PR after that without the test, because the coverage will decrease, what do you think?
(I'm starting to work on the test dataset and test now, it will be in another PR)

File: ivadomed/loader/utils.py
Comment (charleygros) [inline ivadomed/loader/utils.py:580]: :100:

**Comment by charleygros** [comment]
> Thank you @charleygros for your feedback. I added the new params in the config file. I also added the key "extensions" to the default config and modify how the filtering is done with the extensions in the function. That way, "extensions" is not mandatory for MRI (as it was before), and is necessary only if a specific dataset contains more than one possible file extensions (e.g. microscopy with png and ometiff). Could you review the changes? Thanks!

I had a look at it: looks great! Many thanks for documenting this as well!

> I was wondering if it is OK to merge the PR after that without the test, because the coverage will decrease, what do you think?
> (I'm starting to work on the test dataset and test now, it will be in another PR)

Oh yeah, sure, makes complete sense! Let's first merge this PR, then introduce the testing dataset to properly test it. Cheers!

**Pull request merged**
